kafka-client/src/main/java/dev/responsive/kafka/internal/utils/Utils.java

```diff
@@ -122,4 +136,53 @@ public static String extractThreadIdFromThreadName(final String threadName) {
     return threadName;
   }
 
+  public static <K, V> byte[] getValueWithTtl(
+      final Bytes key,
+      final long streamTimeMs,
+      final TtlResolver<K, V> ttlResolver,
+      final Function<Long, ValueAndTimestamp<byte[]>> fetchRecordForMinTs
+  ) {
+    long minValidTsFromKey = 0L;
+    if (ttlResolver.canComputeWithoutValue()) {
+      final TtlDuration ttl = ttlResolver.resolveTtl(key, null);
+      if (ttl.isFinite()) {
+        minValidTsFromKey = streamTimeMs - ttl.toMillis();
+      }
+    }
+
+    final ValueAndTimestamp<byte[]> result = fetchRecordForMinTs.apply(minValidTsFromKey);
+    if (result == null) {
+      return null;
+    }
+
+    if (!ttlResolver.canComputeWithoutValue()) {
+      final TtlDuration ttl = ttlResolver.resolveTtl(key, result.value());
+      if (ttl.isFinite()) {
+        final long minValidTsFromValue = streamTimeMs - ttl.toMillis();
+        final long recordTs = result.timestamp();
+        if (recordTs < minValidTsFromValue) {
+          return null;
+        }
+      }
+    }
+    return result.value();
+  }
+
+  public static <D> byte[] serialize(final D data, final Serde<D> serde) {
+    return serde.serializer().serialize("ignored", data);
+  }
+
+  public static Bytes serializedKey(final String key) {
```

Comment:
what are these used for?

Reply:
tests...guess these should be in the test utils, I'll move them